File: scripts/lightsensors.py
#!/usr/bin/env python
#encoding: utf8                 #長くてコメントが増えそうなので入れました
import sys, rospy
from raspimouse_ros_2.msg import LightSensorValues

def get_freq():    #この関数を追加
    f = rospy.get_param('lightsensors_freq',10)
    try:
        if f <= 0.0:
            raise Exception()
    except:
        rospy.logerr("value error: lightsensors_freq")
        sys.exit(1)

    return f

if __name__ == '__main__':
    devfile = '/dev/rtlightsensor0'
    rospy.init_node('lightsensors')
    pub = rospy.Publisher('lightsensors', LightSensorValues, queue_size=1)

    freq = get_freq()        #追加
    rate = rospy.Rate(freq)  #rate = rospy.Rate(10)から書き換え
    while not rospy.is_shutdown():
        try:
            with open(devfile,'r') as f:
                data = f.readline().split()
                data = [ int(e) for e in data ]
                d = LightSensorValues()
                d.right_forward = data[0]
                d.right_side = data[1]
                d.left_side = data[2]
                d.left_forward = data[3]
                d.sum_all = sum(data)
                d.sum_forward = data[0] + data[3]
                pub.publish(d)
        except IOError:
            rospy.logerr("cannot write to " + devfile)

        f = get_freq()                 #ここから4行追加
        if f != freq:
            freq = f
            rate = rospy.Rate(freq)    #ここまで
    
        rate.sleep()

File: src/lightsensors.cpp
#include "ros/ros.h"
#include <ros/package.h> 
#include "raspimouse_ros_2/LightSensorValues.h"
#include <fstream>
using namespace ros;

int getFrequency(int old, NodeHandle *n)
{
	int f;
	if(n->getParam("frequency",f) and f > 0)
		return f;

	return old;
}

int main(int argc, char **argv)
{
	init(argc,argv,"lightsensors");
	NodeHandle n("~");

	Publisher pub = n.advertise<raspimouse_ros_2::LightSensorValues>("/lightsensors", 5);

	int freq = 10;

	ros::Rate loop_rate(freq);
	raspimouse_ros_2::LightSensorValues msg;

	unsigned int c = 0;
	while(ok()){
		if(c++ % freq == 0){ // check the parapeter every 1[s]
			unsigned int old = freq;
			freq = getFrequency(freq,&n);
			if(old != freq){
				loop_rate = ros::Rate(freq);
				ROS_INFO("Lightsensor frequency: %d", freq);
			}
		}

		std::ifstream ifs("/dev/rtlightsensor0");
		ifs >> msg.right_forward >> msg.right_side
			>> msg.left_side >> msg.left_forward;

		msg.sum_forward = msg.left_forward + msg.right_forward;
		msg.sum_all = msg.sum_forward + msg.left_side + msg.right_side;

		pub.publish(msg);
		spinOnce();
		loop_rate.sleep();
	}
	exit(0);
}

